networking/libsnnet/cnci.go

```diff
@@ -462,6 +471,174 @@ func (cnci *Cnci) addSubnetToTopology(bridge *Bridge, gre *GreTunEP, brInfo **br
 	return
 }
 
+// UpdateNeighbors will create a point to multipoint gre tunnel between
+// all the CNCIs for this tenant.
+func (cnci *Cnci) UpdateNeighbors(neighbors []Neighbor) error {
+	var tun *GreTunEP
+
+	// this must be done first
+	for _, n := range neighbors {
+		if n.PhysicalIP == cnci.ComputeAddr[0].IPNet.IP.String() {
+			IP := net.ParseIP(n.PhysicalIP)
+			if IP == nil {
+				return fmt.Errorf("Unable to parse local physical IP address")
+			}
+			var err error
+
+			tun, err = newGreTunEP("cncitun", IP, n.TunnelID)
+			if err != nil {
+				return err
+			}
+
+			// see if the device already exists
+			err = tun.getDevice()
+			if err != nil {
+				if err = tun.create(); err != nil {
+					return err
+				}
+				if err = tun.enable(); err != nil {
+					return err
+				}
+			}
+
+			// see if my address already exists
+			addrs, err := netlink.AddrList(tun.Link, netlink.FAMILY_V4)
+			if err != nil {
+				return err
+			}
+
+			// XXX: hardcode netmask to 18 for now.
+			addr, err := netlink.ParseAddr(fmt.Sprintf("%s/%d", n.TunnelIP, 18))
+			if err != nil {
+				return err
+			}
+
+			var added bool
+			for _, a := range addrs {
+				if addr.Equal(a) {
+					added = true
+				} else {
+					// we should remove this addr.
+					// we can do this because there should
+					// be only one.
+					err = netlink.AddrDel(tun.Link, &a)
+					if err != nil {
+						return err
+					}
+				}
+			}
+			if !added {
+				err = netlink.AddrAdd(tun.Link, addr)
+				if err != nil {
+					return err
+				}
+			}
+			break
+		}
+	}
+	neighs, err := netlink.NeighList(tun.Link.Index, netlink.FAMILY_V4)
+	if err != nil {
+		return err
+	}
+
+	var updated []netlink.Neigh
+	for _, n := range neighbors {
+		if n.PhysicalIP == cnci.ComputeAddr[0].IPNet.IP.String() {
+			continue
+		}
+
+		neigh := netlink.Neigh{
+			IP:        net.ParseIP(n.TunnelIP),
+			LLIPAddr:  net.ParseIP(n.PhysicalIP),
+			LinkIndex: tun.Link.Index,
+			State:     netlink.NUD_PERMANENT,
+		}
+		updated = append(updated, neigh)
+
+		var exists bool
+		// see if this already exists
+		for _, neighbor := range neighs {
+			if strings.Compare(neighbor.String(), neigh.String()) == 0 {
```

Comment:
Is this is the right comparison. We are comparing full equality.

However what we should be checking is if the IP is reachable via a different LLIPAddr.

So I think we should be checking if IP matches. We should not have the same IP reachable via two different LLAddrs.

So if IP matchs and we have a full match the config has not changed. And we skip ahead.

However if IP matches but LLIPAddr does not match, we should delete the old entry and add a new one.

Reply:
ok - I'll fix this.